### PR TITLE
[DNM] Ports Collar ID's.

### PR DESCRIPTION
**Pull request by Kevak**
Collars can now hold ID's.
ID's in collars count for access when checking access.
ID's can be removed from collars by alt clicking similar to PDA's.
Shock collars also can hold ID's, with slightly tweaked flufftext.
Game checks for neckslot access as well as the normal handslots and IDslot.

Everything has been tested. Report identified bugs on the cit git.

As a side note. It would be 'possible' for me make PAI's fit in collars. It seems moderately silly but it is a thing I can do. Lemme know if people want that. Could prolly even give em a zap button for shock collars.

:cl: Kevak/Neri
add: Collars can now hold ID's. Stored ID's count for access.
add: ID's can be removed by collars via altclick, same as how PDA's work.
fix: Cleaned up Shockcollar code slightly.
/:cl:

Why? Why not? It's another one of those lil gimmicks that very few people will use but it'll be nice to have as an option. It's neat and it's there.

### CODE REVIEW

File: code/game/objects/items/devices/radio/electropack.dm [hunk 1]
```diff
@@ -222,3 +223,82 @@ Code:
...
 			return
 
+/obj/item/device/electropack/shockcollar/proc/remove_id()
+	if (id)
+		if (ismob(loc))
+			var/mob/M = loc
+			M.put_in_hands(id)
+			to_chat(usr, "<span class='notice'>You remove the ID from the [name].</span>")
+		else
+			id.loc = get_turf(src)
```
Comment (silicons) [inline code/game/objects/items/devices/radio/electropack.dm]: forceMove

File: code/game/objects/items/devices/radio/electropack.dm [hunk 2]
```diff
@@ -222,3 +223,82 @@ Code:
...
+		id = null
+
+/obj/item/device/electropack/shockcollar/verb/verb_remove_id()
```
Comment (silicons) [inline code/game/objects/items/devices/radio/electropack.dm]: Upstream has taught me that verbs are horrid but this is passable I guess.

File: code/game/objects/items/devices/radio/electropack.dm [hunk 3]
```diff
@@ -222,3 +223,82 @@ Code:
...
+		if(id)
+			remove_id()
+		else
+			to_chat(usr, "<span class='warning'>This collar does not have an ID in it!</span>")
+
+/obj/item/device/electropack/shockcollar/proc/id_check(mob/user, obj/item/card/id/I)
+	if(!I)
+		if(id)
+			remove_id()
+			return 1
```
Comment (silicons) [inline code/game/objects/items/devices/radio/electropack.dm]: TRUE

File: code/game/objects/items/devices/radio/electropack.dm [hunk 4]
```diff
@@ -222,3 +223,82 @@ Code:
...
+		else
+			var/obj/item/card/id/C = user.get_active_held_item()
+			if(istype(C))
+				I = C
+
+	if(I)
+		if(!user.transferItemToLoc(I, src))
+			return 0
```
Comment (silicons) [inline code/game/objects/items/devices/radio/electropack.dm]: FALSE

File: code/game/objects/items/devices/radio/electropack.dm [hunk 5]
```diff
@@ -222,3 +223,82 @@ Code:
...
+		var/obj/old_id = id
+		id = I
+		if(old_id)
+			user.put_in_hands(old_id)
+		update_icon()
+	return 1
```
Comment (silicons) [inline code/game/objects/items/devices/radio/electropack.dm]: TRUE

File: code/game/objects/items/devices/radio/electropack.dm [hunk 6]
```diff
@@ -222,3 +223,82 @@ Code:
...
+			if(istype(C))
+				I = C
+
+	if(I)
+		if(!user.transferItemToLoc(I, src))
+			return 0
+		var/obj/old_id = id
+		id = I
+		if(old_id)
+			user.put_in_hands(old_id)
```
Comment (silicons) [inline code/game/objects/items/devices/radio/electropack.dm]: What happens if this fails?

File: code/game/objects/items/devices/radio/electropack.dm [hunk 7]
```diff
@@ -222,3 +223,82 @@ Code:
...
+	if(istype(C, /obj/item/card/id))
+		var/obj/item/card/id/idcard = C
+		if(((src in user.contents) || (isturf(loc) && in_range(src, user))) && (C in user.contents))
+			if(!id_check(user, idcard))
+				to_chat(user, "<span class='warning'>There is already \a [id] slotted into \the [src]</span>")
+				return
+			to_chat(user, "<span class='notice'>\The [id] slots into \the [src] snugly.</span>")
+			return
+	if(!user.transferItemToLoc(C, src))
+		return
```
Comment (silicons) [inline code/game/objects/items/devices/radio/electropack.dm]: Two?

File: code/modules/clothing/neck/neck.dm [hunk 8]
```diff
@@ -148,11 +148,92 @@
...
 	name = "[initial(name)] - [tagname]"
 
+/obj/item/clothing/neck/petcollar/proc/remove_id()
+	if (id)
+		if (ismob(loc))
+			var/mob/M = loc
+			M.put_in_hands(id)
+			to_chat(usr, "<span class='notice'>You remove the ID from the [name].</span>")
+		else
+			id.loc = get_turf(src)
```
Comment (silicons) [inline code/modules/clothing/neck/neck.dm]: forceMove

File: code/modules/clothing/neck/neck.dm [hunk 9]
```diff
@@ -148,11 +148,92 @@
...
 
 /obj/item/clothing/neck/petcollar/attack_self(mob/user)
 	tagname = copytext(sanitize(input(user, "Would you like to change the name on the tag?", "Name your new pet", "Spot") as null|text),1,MAX_NAME_LEN)
 	name = "[initial(name)] - [tagname]"
 
+/obj/item/clothing/neck/petcollar/proc/remove_id()
+	if (id)
+		if (ismob(loc))
+			var/mob/M = loc
+			M.put_in_hands(id)
```
Comment (silicons) [inline code/modules/clothing/neck/neck.dm]: What if this fails?

File: code/modules/clothing/neck/neck.dm [hunk 10]
```diff
@@ -148,11 +148,92 @@
 	icon_state = "petcollar"
 	item_color = "petcollar"
 	var/tagname = null
+	var/obj/item/card/id/id = null
```
Comment (silicons) [inline code/modules/clothing/neck/neck.dm]: No = null it's implied.

File: code/modules/clothing/neck/neck.dm [hunk 11]
```diff
@@ -148,11 +148,92 @@
...
+		if(id)
+			remove_id()
+		else
+			to_chat(usr, "<span class='warning'>This collar does not have an ID in it!</span>")
+
+/obj/item/clothing/neck/petcollar/proc/id_check(mob/user, obj/item/card/id/I)
+	if(!I)
+		if(id)
+			remove_id()
+			return 1
```
Comment (silicons) [inline code/modules/clothing/neck/neck.dm]: TRUE

File: code/modules/clothing/neck/neck.dm [hunk 12]
```diff
@@ -148,11 +148,92 @@
...
+		else
+			var/obj/item/card/id/C = user.get_active_held_item()
+			if(istype(C))
+				I = C
+
+	if(I)
+		if(!user.transferItemToLoc(I, src))
+			return 0
```
Comment (silicons) [inline code/modules/clothing/neck/neck.dm]: FALSE

File: code/modules/clothing/neck/neck.dm [hunk 13]
```diff
@@ -148,11 +148,92 @@
...
+		var/obj/old_id = id
+		id = I
+		if(old_id)
+			user.put_in_hands(old_id)
+		update_icon()
+	return 1
```
Comment (silicons) [inline code/modules/clothing/neck/neck.dm]: TRUE

File: code/modules/clothing/neck/neck.dm [hunk 14]
```diff
@@ -148,11 +148,92 @@
...
+	if(istype(C, /obj/item/card/id))
+		var/obj/item/card/id/idcard = C
+		if(((src in user.contents) || (isturf(loc) && in_range(src, user))) && (C in user.contents))
+			if(!id_check(user, idcard))
+				to_chat(user, "<span class='warning'>There is already \a [id] clipped onto \the [src]</span>")
+				return
+			to_chat(user, "<span class='notice'>\The [id] clips onto \the [src] snugly.</span>")
+			return
+	if(!user.transferItemToLoc(C, src))
+		return
```
Comment (silicons) [inline code/modules/clothing/neck/neck.dm]: Two?

File: code/modules/jobs/access.dm [hunk 15]
```diff
@@ -19,7 +19,7 @@
 	else if(ishuman(M))
 		var/mob/living/carbon/human/H = M
 		//if they are holding or wearing a card that has access, that works
-		if(check_access(H.get_active_held_item()) || src.check_access(H.wear_id))
+		if(check_access(H.get_active_held_item()) || src.check_access(H.wear_id) || src.check_access(H.wear_neck))
```
Comment (silicons) [inline code/modules/jobs/access.dm]: No .src

**Comment by silicons** [comment]
This is a preliminary review . I'm assuming you fully tested this too.

File: code/game/objects/items/devices/radio/electropack.dm (hunk 2 above)
Comment (Kevak) [reply code/game/objects/items/devices/radio/electropack.dm]: It worked in testing. I could prolly nix it though. Not like it's essential. Will note that PDA uses the same thing.

File: code/modules/clothing/neck/neck.dm (hunk 9 above)
Comment (Kevak) [reply code/modules/clothing/neck/neck.dm]: No idea! It didn't in testing but I should prolly add an else in or something.

File: code/game/objects/items/devices/radio/electropack.dm (hunk 7 above)
Comment (Kevak) [reply code/game/objects/items/devices/radio/electropack.dm]: Shock collars are not children of petcollars, they're children of electropacks due to electropacks being the more complex code. As such, need two, one for /petcollar and one for /electropack/shockcollar

File: code/modules/clothing/neck/neck.dm (hunk 14 above)
Comment (Kevak) [reply code/modules/clothing/neck/neck.dm]: Shock collars are not children of petcollars, they're children of electropacks due to electropacks being the more complex code. As such, need two, one for /petcollar and one for /electropack/shockcollar

**Comment by Kevak** [comment]
I did indeed fully test. Will make edits based on feedback.

File: code/modules/clothing/neck/neck.dm (hunk 9 above)
Comment (Kevak) [reply code/modules/clothing/neck/neck.dm]: Main reason that's there by the way is cause it's yoinked from PDAcode. PDAcode has a lotta old shit.

File: code/modules/clothing/neck/neck.dm (hunk 14 above)
Comment (silicons) [reply code/modules/clothing/neck/neck.dm]: I mean two returns?

File: code/game/objects/items/devices/radio/electropack.dm (hunk 7 above)
Comment (silicons) [reply code/game/objects/items/devices/radio/electropack.dm]: I mean two returns?

Comment (Kevak) [reply code/game/objects/items/devices/radio/electropack.dm]: Ah. Probably not. Will change that.

File: code/modules/clothing/neck/neck.dm (hunk 14 above)
Comment (Kevak) [reply code/modules/clothing/neck/neck.dm]: Prolly not. Will change.

**Comment by Kevak** [comment]
I think this is prolly ready for review again. Anything else need fixing?

File: code/game/objects/items/devices/radio/electropack.dm [hunk 16]
```diff
@@ -222,3 +223,81 @@ Code:
 			onclose(user, "radio")
 			return
 
+/obj/item/device/electropack/shockcollar/proc/remove_id()
+	if (id)
```
Comment (silicons) [inline code/game/objects/items/devices/radio/electropack.dm]: id && ismob(loc)

File: code/game/objects/items/devices/radio/electropack.dm [hunk 17]
```diff
@@ -222,3 +223,81 @@ Code:
...
+		if (ismob(loc))
+			var/mob/M = loc
+			M.put_in_hands(id)
```
Comment (silicons) [inline code/game/objects/items/devices/radio/electropack.dm]: what if this fails it should drop to ground or something

File: code/game/objects/items/devices/radio/electropack.dm [hunk 18]
```diff
@@ -222,3 +223,81 @@ Code:
...
+			to_chat(usr, "<span class='notice'>You remove the ID from the [name].</span>")
+		else
+			id.forceMove(get_turf(src))
```
Comment (silicons) [inline code/game/objects/items/devices/radio/electropack.dm]: forceMove(drop_location())

File: code/modules/clothing/neck/neck.dm [hunk 19]
```diff
@@ -148,11 +148,91 @@
...
+			return TRUE
+		else
+			var/obj/item/card/id/C = user.get_active_held_item()
+			if(istype(C))
+				I = C
+
+	if(I)
+		if(!user.transferItemToLoc(I, src))
+			return FALSE
+		var/obj/old_id = id
```
Comment (silicons) [inline code/modules/clothing/neck/neck.dm:179]: why not eject old and then set new so you don't need an extra var?

File: code/modules/clothing/neck/neck.dm [hunk 20]
```diff
@@ -148,11 +148,91 @@
...
+		id = I
+		if(old_id)
+			user.put_in_hands(old_id)
```
Comment (silicons) [inline code/modules/clothing/neck/neck.dm:182]: what if this fails?

File: code/modules/clothing/neck/neck.dm (hunk 19 above)
Comment (Kevak) [reply code/modules/clothing/neck/neck.dm:179]: Hooonestly I literally just copied that from how PDA's work. Soo.

File: code/modules/clothing/neck/neck.dm (hunk 20 above)
Comment (Kevak) [reply code/modules/clothing/neck/neck.dm:182]: Think I just fixed.

File: code/game/objects/items/devices/radio/electropack.dm (hunk 17 above)
Comment (Kevak) [reply code/game/objects/items/devices/radio/electropack.dm]: Fixed I think.

File: code/game/objects/items/devices/radio/electropack.dm (hunk 18 above)
Comment (Kevak) [reply code/game/objects/items/devices/radio/electropack.dm]: Fixed. I think.

File: code/game/objects/items/devices/radio/electropack.dm (hunk 16 above)
Comment (Kevak) [reply code/game/objects/items/devices/radio/electropack.dm]: Fixed I think.

File: code/game/objects/items/devices/radio/electropack.dm [hunk 21]
```diff
@@ -158,6 +158,7 @@ Code:
 	equip_delay_other = 60
 	materials = list(MAT_METAL=5000, MAT_GLASS=2000)
 	var/tagname = null
+	var/obj/item/card/id/id = null
```
Comment (silicons) [inline code/game/objects/items/devices/radio/electropack.dm]: no = null it's implied

File: code/game/objects/items/devices/radio/electropack.dm [hunk 22]
```diff
@@ -222,3 +223,82 @@ Code:
 			onclose(user, "radio")
 			return
 
+/obj/item/device/electropack/shockcollar/proc/remove_id()
+	if (id && ismob(loc))
+		var/mob/M = loc
+		M.put_in_hands(id)
```
Comment (silicons) [inline code/game/objects/items/devices/radio/electropack.dm]: if(!M.put_in_hands(id))
    id.forceMove(drop_location())

File: code/game/objects/items/devices/radio/electropack.dm [hunk 23]
```diff
@@ -222,3 +223,82 @@ Code:
...
+			return TRUE
+		else
+			var/obj/item/card/id/C = user.get_active_held_item()
+			if(istype(C))
+				I = C
+
+	if(I)
+		if(!user.transferItemToLoc(I, src))
+			return FALSE
+		var/obj/old_id = id
```
Comment (kevinz000) [inline code/game/objects/items/devices/radio/electropack.dm]: replace this entire block with
`
if(old_id)
    if(!user.put_in_hands(old_id))
        old_id.forceMove(drop_location())
    id = I
`

Comment (kevinz000) [reply code/game/objects/items/devices/radio/electropack.dm]: formatting's fucked up by github but you get the point

File: code/game/objects/items/devices/radio/electropack.dm [hunk 24]
```diff
@@ -222,3 +223,82 @@ Code:
...
+		id = I
+		if(old_id)
+			user.put_in_hands(old_id)
```
Comment (silicons) [inline code/game/objects/items/devices/radio/electropack.dm:251]: this has to do something if it fails or else you just trapped an id in this forever.

File: code/modules/clothing/neck/neck.dm (hunk 20 above)
Comment (silicons) [reply code/modules/clothing/neck/neck.dm:182]: no it still doesn't move old_id anywhere if it can't put in hands

**Comment by Kevak** [comment]
Wait fucking what. Why is it trying to do stuff to boxstation.dmm

**Comment by Kevak** [comment]
Also made a new bug. Looking into it.

File: code/game/objects/items/devices/radio/electropack.dm (hunk 24 above)
Comment (Kevak) [reply code/game/objects/items/devices/radio/electropack.dm:251]: Fixing.

File: code/modules/clothing/neck/neck.dm (hunk 20 above)
Comment (Kevak) [reply code/modules/clothing/neck/neck.dm:182]: Fixing.

**Comment by Kevak** [comment]
Aaugh fuck it. It's late and I need sleep. Will fix bug tomorrow.

If ye happen to skim through, assistance in locating the problem would be appreciated. Am a bit stumped.
Bug is: If collar empty, and ID is used on collar, ID moves to floor below usermob, Collar behaves as if it has access in accordance to the card. If you apply a second card, the first card teleports to hand and the second card slots in without issue. Issue is _Probably_ something to do with the forcemove definers? Not quite sure.

**Comment by LetterJay** [comment]
Is this merge-ready?

**Comment by Kevak** [comment]
No. Been fixing bones. Flag DNM and I'll ping ya when it's merge ready.

**Comment by LetterJay** [comment]
For future reference, you can put [DNM] in the title to have it autotag.

**Comment by Kevak** [comment]
Good to know.

**Comment by LetterJay** [comment]
This is also stale.

**Pull request closed without merging**